[![CircleCI Build Status](https://circleci.com/gh/circleci/circleci-docs.svg?style=shield)](https://circleci.com/gh/BU-ISCIII/bacterial_wgs_training) [![License: GPL v3](https://img.shields.io/badge/License-GPL%20v3-blue.svg)](https://www.gnu.org/licenses/gpl-3.0) [![CircleCi Community](https://img.shields.io/badge/community-CircleCI%20Discuss-343434.svg)](https://discuss.circleci.com) [![Nextflow version](https://img.shields.io/badge/nextflow->0.29.0-green.svg)](http://nextflow.io) [![Scif](https://img.shields.io/badge/Filesystem-Scientific-brightgreen.svg)](https://sci-f.github.io)

# Bacterial WGS training
In this training course you will find theory and practice material for introducing yourself to wgs analysis for bacterial, including outbreak investigation. [Here you will find the agenda.](slides/20221023_4ED_curso_SeqGenBac_agenda.pdf)

The material includes slides with theory concepts and a bunch of practical exercises using nextflow and singularity, focusing on the interpretation of results.

## Slides
### Day 1
- **Talk 1 and 2:** [Massive sequencing of bacterial genomes. State-of-the-art. & Bacterial genomes sequencing. Applications.](slides/talk1/20221024_4ED_curso_SeqGenBac_session1.1-2_Introduccion_ICuesta.pdf)
- **Talk 3:** [Linux environment review.](slides/talk2/curso_SeqGenBac_session1.2_linux.pdf)
- [**Exercise 0**](exercises/00_SetUp.md) -- [Download pdf](exercises/00_Setup.pdf)
- [**Exercise 1**](exercises/01_LinuxBasicCommands.md) -- [Download pdf](exercises/01_LinuxBasicCommands.pdf)

### Day 2
- **Talk 4:** [The computing revolution in Biosciences. Nextflow and Singularity introduction.](slides/talk3/curso_SeqGenBac_ChangingComputingParadigm.pdf)
- [**Exercise 2**](exercises/02_NextflowSingularity.md) -- [Download pdf](exercises/02_LinuxNextflowSingularity.pdf)

- **Talk 5:** [Quality analysis and control of HTS data](slides/talk5/curso_SeqGenBac_session2.2_quality_assesment.pdf)

- **Talk 6:** [Bacterial genomes assembly](slides/talk6/curso_SeqGenBac_session2.3_assembly.pdf)

- [**Exercise 3**](exercises/02_QualityAndAssembly.md) -- [Download pdf](exercises/02_QualityAndAssembly.pdf)

### Day 3
- **Talk 7:** [Mapping against reference genome and Variant Calling.](slides/talk7/curso_SeqGenBac_session3.1_MappingAndVariantCalling.pdf)
- **Talk 8:** [SNP matrix and phylogenetics.](slides/talk8/curso_SeqGenBac_session3.2_SNPMatrixAndPhylogenetics.pdf)

- [**Exercise 4**](exercises/03_outbreakSNP.md) -- [Download pdf](exercises/03_outbreakSNP.pdf)

### Day 4
- **Talk 9:** [Typing based on allelic profile or gene-by-gene](slides/talk9/curso_SeqGenBac_session4.1_tipificacion-gen-by-gene_ICuesta.pdf)
- **Talk 10:** [Gene-by-gene WGS analysis](slides/talk10/curso_SeqGenBac_session4.2_GeneByGenevsSNPs_v2.pdf)

- [**Exercise 5**](exercises/04_outbreakcgMLST.md) -- [Download pdf](exercises/04_outbreakcgMLST.pdf)

### Day 5
- **Talk 11:** [Sequence annotation](slides/talk11/curso_SeqGenBac_session5.1_annotation.pdf)
- **Talk 12:** [Genome characterization, Resistance and Virulence genes]
- [**Exercise 6**](exercises/05_annotation.md) -- [Download pdf](exercises/05_annotation.pdf)
